code/service/src/main/scala/com/namely/chiefofstate/ServiceBootstrapper.scala

```diff
@@ -0,0 +1,188 @@
+import com.typesafe.config.Config
+import io.grpc._
+import io.grpc.netty.NettyServerBuilder
+import io.opentelemetry.api.GlobalOpenTelemetry
+import io.opentelemetry.instrumentation.grpc.v1_5.GrpcTracing
+
+import java.net.InetSocketAddress
+import scala.concurrent.ExecutionContext
+import scala.sys.ShutdownHookThread
+
+/**
+ * This helps setup the required engines needed to smoothly run the ChiefOfState sevice.
+ * The following engines are started on boot.
+ * <ul>
+ *   <li> the akka cluster system
+ *   <li> the akka cluster sharding engine
+ *   <li> the akka cluster management
+ *   <li> loads the various ChiefOfState plugins
+ *   <li> run the required schemas and migration needed
+ *   <li> the telemetry tools and the various gRPC interceptors
+ *   <li> the gRPC service
+ * </ul>
+ */
+object ServiceBootstrapper {
+  // define the bootstrap protocol
+  sealed trait StartCommand extends CborSerializable
+  final case class StartMigration(replyTo: ActorRef[StartMigrationReply]) extends StartCommand
+  final case class StartMigrationReply() extends StartCommand
+
+  def apply(config: Config): Behavior[StartCommand] = Behaviors
+    .setup[StartCommand] { context =>
+      // get the  COS config
+      val cosConfig: CosConfig = CosConfig(config)
+
+      Behaviors.receiveMessage[StartCommand] {
+        case StartMigrationReply() =>
+          // start the telemetry tools and register global tracer
+          TelemetryTools(cosConfig).start()
+
+          // We only proceed when the data stores and various migrations are done successfully.
+          log.info("Journal and snapshot store created successfully. About to start...")
+
+          val channel: ManagedChannel =
+            NettyHelper
+              .builder(
+                cosConfig.writeSideConfig.host,
+                cosConfig.writeSideConfig.port,
+                cosConfig.writeSideConfig.useTls
+              )
+              .build()
+
+          val grpcClientInterceptors: Seq[ClientInterceptor] = Seq(
+            GrpcTracing.create(GlobalOpenTelemetry.get()).newClientInterceptor(),
+            new StatusClientInterceptor()
+          )
+
+          val writeHandler: WriteSideHandlerServiceBlockingStub = new WriteSideHandlerServiceBlockingStub(channel)
+            .withInterceptors(grpcClientInterceptors: _*)
+
+          val remoteCommandHandler: RemoteCommandHandler = RemoteCommandHandler(cosConfig.grpcConfig, writeHandler)
+          val remoteEventHandler: RemoteEventHandler = RemoteEventHandler(cosConfig.grpcConfig, writeHandler)
+
+          // instance of eventsAndStatesProtoValidation
+          val eventsAndStateProtoValidation: ProtosValidator = ProtosValidator(
+            cosConfig.writeSideConfig
+          )
+
+          // initialize the sharding extension
+          val sharding: ClusterSharding = ClusterSharding(context.system)
+
+          // initialize the shard region
+          sharding.init(
```

Comment:
lets add a strategy here to wait for the migrations to be complete before starting the persistence & projections. I'm imagining something really simple, like polling the `cos_migrations` table until we reach the max, current version.